strawberryfields/parameters.py

```diff
@@ -19,335 +19,334 @@
 
 .. currentmodule:: strawberryfields.parameters
 
-The :class:`Parameter` class encapsulates a parameter passed to the
+The classes in this module encapsulate parameters passed to the
 quantum operations represented by :class:`.Operation`.
-The parameter objects can represent a
-
-* number
-* NumPy array
-* value measured from the quantum register (:class:`.RegRefTransform`)
-* TensorFlow object (:class:`tf.Variable` or :code:`tf.placeholder`)
-
-
-The normal lifecycle of an Operation object and its associated Parameter instances is as follows:
+There are three basic types of parameters:
+
+1. Numerical parameters (bound and fixed): An immediate, immutable numerical object
+   (float, complex, int, array).
+   NOTE: currently implemented as-is, not encapsulated in a class. This can be changed if necessary,
+   in which case most of the top-level functions in this module would become Parameter methods.
+2. Measured parameters (bound but not fixed): Certain quantum circuits/protocols require that
+   Operations can be conditioned on measurement results obtained during the execution of the
+   circuit. In this case the parameter value is not known/fixed until the measurement is made
+   (or simulated). Represented by :class:`MeasuredParameter` instances.
+   Constructed from the :class:`RegRef` instance storing the measurement
+   result using the :meth:`RegRef.par` method.
+3. Free parameters (not bound nor fixed): A *parametrized circuit template* is a circuit that
+   depends on a number of unbound (free) parameters. These parameters need to be bound to fixed
+   numerical values before the circuit can be executed on a hardware quantum device or a numeric
+   simulator. Represented by :class:`FreeParameter` instances.
+   Simulators with symbolic capability can accept a parametrized circuit as input (and should
+   return symbolic expressions representing the measurement results, with the same free parameters,
+   as output).
+   Free parameters belong to a single :class:`Program` instance, are constructed using the
+   :meth:`Program.args` method, and are bound using :meth:`Program.bind_args`.
```

Comment:
This is a completely new feature in this PR, right? Are there details on how it is accessible from the frontend once there are compatible backends? Or even, what a compatible backend must do?

Reply:
This is a frontend-only feature, `Engine.run` binds all the free parameters. However, they could be bound to e.g. TensorFlow objects which are then passed to the backend, in which case the backend must be able to deal with them.